src/client/dfuse/dfuse_core.c

```diff
@@ -691,10 +691,16 @@ dfuse_cont_open_by_label(struct dfuse_info *dfuse_info, struct dfuse_pool *dfp,
 
 	DFUSE_TRA_UP(dfc, dfp, "dfc");
 
-	rc = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RW, &dfc->dfs_coh, &c_info, NULL);
-	if (rc == -DER_NO_PERM) {
+	if (dfuse_info->di_read_only) {
 		dfs_flags = O_RDONLY;
 		rc = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RO, &dfc->dfs_coh, &c_info, NULL);
+	} else {
+		rc = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RW, &dfc->dfs_coh, &c_info, NULL);
+		if (rc == -DER_NO_PERM) {
+			dfs_flags = O_RDONLY;
+			rc        = daos_cont_open(dfp->dfp_poh, label, DAOS_COO_RO, &dfc->dfs_coh,
+						   &c_info, NULL);
+		}
```

Comment:
i feel like it's better to fail in this case rather than retry with RO if user did not request RO.

Reply:
This is existing code but for dfuse then in many cases the user doesn't explicitly request either mode, they just see a path and try to access it.  Failing back to RO was something that came up early on in the multi-user testing as something that is likely to happen, I'll see if I can find the ticket.

Even this PR doesn't allow setting of read-only mode on a per container basis, it's more saying "never write" rather than read-only, in general dfuse just doesn't know if access will be RO or RW so it needs to be flexible.